bob-backend/src/pearl/group.rs

```diff
@@ -569,39 +598,79 @@ impl Group {
         });
     }
 
-    pub async fn delete(&self, key: BobKey) -> Result<u64, Error> {
-        let holders = self.holders.read().await;
+    pub async fn delete(
+        &self,
+        key: BobKey,
+        is_alien: bool,
+        timestamp_config: StartTimestampConfig,
+        force_delete: bool,
+    ) -> Result<u64, Error> {
+        let is_empty = self.holders.read().await.len() == 0;
```

Comment:
Most of the time `is_empty` will be `false`. But to check that, you acquire lock, which is ineffective. It is better to assume, that `holders` collection is not empty and write the code with that assumption:
```
{
let lock = self.holders.read().await;
if lock.len() != 0 {
    return  self.delete_in_holders(holders, key, is_alien, force_delete).await;
}
}
self.get_actual_holder(get_current_timestamp(), timestamp_config).await?;
{
let lock = self.holders.read().await;
return self.delete_in_holders(holders, key, is_alien, force_delete).await;
}
```

Reply:
1. Second time we don't need holders lock, since actual holder is already there.